concurrent/src/test/java/io/airlift/concurrent/TestBoundedThreadPool.java

```diff
@@ -0,0 +1,104 @@
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+
+public class TestBoundedThreadPool
+{
+    private ExecutorService executor;
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+            throws Exception
+    {
+        executor.shutdownNow();
+    }
+
+    @Test
+    public void testCounter()
+            throws Exception
+    {
+        executor = newBoundedThreadPool(1, threadsNamed("test-%s")); // Enforce single thread
+
+        int totalTasks = 100_000;
+        AtomicInteger counter = new AtomicInteger();
+        CountDownLatch startLatch = new CountDownLatch(1);
+        CountDownLatch completeLatch = new CountDownLatch(totalTasks);
+
+        for (int i = 0; i < totalTasks; i++) {
+            executor.execute(() -> {
+                try {
+                    awaitUninterruptibly(startLatch); // Wait for the go signal
+
+                    // Intentional distinct read and write calls
+                    int initialCount = counter.get();
+                    counter.set(initialCount + 1);
+                }
+                finally {
+                    completeLatch.countDown();
+                }
+            });
+        }
+
+        startLatch.countDown(); // Signal go for threads
+        awaitUninterruptibly(completeLatch, 1, MINUTES); // Wait for tasks to complete
+        assertEquals(counter.get(), totalTasks);
+    }
+
+    @Test
+    public void testSingleThreadBound()
+            throws Exception
+    {
+        testBound(1, 100_000);
+    }
+
+    @Test
+    public void testDoubleThreadBound()
+            throws Exception
+    {
+        testBound(2, 100_000);
+    }
+
+    private void testBound(int maxThreads, int totalTasks)
+    {
+        executor = newBoundedThreadPool(maxThreads, threadsNamed("test-%s"));
+
+        AtomicInteger activeThreadCount = new AtomicInteger();
+        CountDownLatch startLatch = new CountDownLatch(1);
+        CountDownLatch completeLatch = new CountDownLatch(totalTasks);
+        AtomicBoolean failed = new AtomicBoolean();
+
+        for (int i = 0; i < totalTasks; i++) {
+            executor.execute(() -> {
+                try {
+                    awaitUninterruptibly(startLatch); // Wait for the go signal
+                    int count = activeThreadCount.incrementAndGet();
+                    if (count < 1 || count > maxThreads) {
+                        failed.set(true);
+                    }
+                    activeThreadCount.decrementAndGet();
+                }
+                finally {
+                    completeLatch.countDown();
+                }
+            });
+        }
+
+        startLatch.countDown(); // Signal go for threads
+        awaitUninterruptibly(completeLatch, 1, MINUTES); // Wait for tasks to complete
```

Comment:
Oh, you're doing it with the countdown latch. Hmm.. the problem with that is that one execution is "lost", the latch fail with a "timeout error". It will take 1 minute for the test to fail, and the error won't be very clear about what happened.